rskj-core/src/main/java/co/rsk/pcc/NativeContract.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * This file is part of RskJ
+ * Copyright (C) 2019 RSK Labs Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package co.rsk.pcc;
+
+import co.rsk.config.RskSystemProperties;
+import co.rsk.core.RskAddress;
+import co.rsk.panic.PanicProcessor;
+import org.bouncycastle.util.encoders.Hex;
+import org.ethereum.core.Block;
+import org.ethereum.core.Repository;
+import org.ethereum.core.Transaction;
+import org.ethereum.db.BlockStore;
+import org.ethereum.db.ReceiptStore;
+import org.ethereum.vm.LogInfo;
+import org.ethereum.vm.PrecompiledContracts;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * This class contains common behavior for a RSK native contract.
+ *
+ * @author Ariel Mendelzon
+ */
+public abstract class NativeContract extends PrecompiledContracts.PrecompiledContract {
+    private static final Logger logger = LoggerFactory.getLogger(NativeContract.class);
+    private static final PanicProcessor panicProcessor = new PanicProcessor();
+
+    private final RskSystemProperties config;
+
+    private ExecutionEnvironment executionEnvironment;
+
+    public NativeContract(RskSystemProperties config, RskAddress contractAddress) {
+        this.contractAddress = contractAddress;
+        this.config = config;
+    }
+
+    public ExecutionEnvironment getExecutionEnvironment() {
+        return executionEnvironment;
+    }
+
+    public abstract List<NativeMethod> getMethods();
+
+    public abstract Optional<NativeMethod> getDefaultMethod();
+
+    public void before() {
```

Comment:
I don't see these methods being used so far. Maybe leave them out while they're not needed?

Reply:
This work is based of what the `Bridge` currently does, which is a uniform setup/teardown of storage on every execution. I wanted to leave that capability on the framework in case someone wants to do something fairly standard out of the box.

Reply:
The Core team follows the YAGNI principle. We have a long-term goal of minimizing maintenance costs, and one of the things we do is avoid (and remove) unused code and features.

By the way, if you were migrating the `Bridge` contract to be a `NativeContract` as part of this PR, these methods would've been great.

Reply:
I understand you as a team use the YAGNI principle and I generally agree with it.

That said, I believe principles themselves are not to be taken at face value. In this case, the code is absolutely bare minimal and has the intention of pointing developers in what is the desired direction on a fairly common pattern arising in native contracts. This sort of "protects" the framework from getting things put and/or changed anywhere in the future with a high risk of causing bugs given that the execution context is highly complex.

If, on the other hand, this actually implied a fair bit of code, with several lines and certain algorithmic complexity, I would definitely agree with you and remove it.